### PR TITLE
tracing: fix zipkin annotation timestamp serialization issue

### LINKED ISSUE #15960: Zipkin trace span timestamp annotation is not properly replaced during serialization
https://github.com/envoyproxy/envoy/pull/15492 adds timestamp annotations to span, but the replacement records is not properly set:

https://github.com/envoyproxy/envoy/blob/110aec2b2fb61c8c9d3e7295b08beb8d46cc01e3/source/extensions/tracers/zipkin/span_buffer.cc#L136

With the code above, it records replacement as key: `"cr":"1618349542333171"`   value: `"cr":1618349542333171`, but the serialized annotation string is like `"annotations":[{"value":"cr","timestamp":"1618349542333171"}]`. This make jaeger trace report fails due to wrong timestamp data type: https://github.com/istio/istio/issues/32026

cc @dio @jamesmulcahy

### PR DESCRIPTION
**Pull request by douglas-reid**
Commit Message: fix handling of timestamps in zipkin annotations / logs
Additional Description: the intended replacement from string to int was not happening as expected. this causes ingestion issues with non-lenient collectors, such as jaeger.  Augmented existing tests to use different timestamps for annotations to catch issue and prevent regression.
Risk Level: Low
Testing: Unit
Docs Changes: None
Release Notes: None
Platform Specific Features: None

Fixes #15960 

Signed-off-by: Douglas Reid <douglas-reid@users.noreply.github.com>

### CODE REVIEW

**Review by bianpengyuan — APPROVED** [review]

**Comment by jamesmulcahy** [comment]
As we updated to v1.18.2 we ran into this issue too. One of our integration tests validates the zipkin integration, but for various reasons, that hadn't been run in conjunction with the original patch.  The test validates the data against the zipkin-go objects, and was running into the same string/int issue.

We've tested this fix this morning and confirmed that the integration test now passes. 

Thanks @douglas-reid.  From a review perspective, you may want to add an entry to the bugfixes section of the release notes?

**Comment by jamesmulcahy** [comment]
I also wonder if we should get this into a v1.18.3 -- others may well run into this too.

**Comment by douglas-reid** [comment]
> As we updated to v1.18.0 we ran into this issue too. One of our integration tests validates the zipkin integration, but for various reasons, that hadn't been run in conjunction with the original patch. The test validates the data against the zipkin-go objects, and was running into the same string/int issue.
> 
> We've tested this fix this morning and confirmed that the integration test now passes.
> 
> Thanks @douglas-reid. From a review perspective, you may want to add an entry to the bugfixes section of the release notes?
> 
> Hat Tip to @drobertduke from my team, who actually did the work/testing that I mentioned above -- I'm just the messenger.

Added to the release notes.

Agree that we should get this backported as well.

**Comment by douglas-reid** [comment]
@dio any chance for a quick review on this PR?  It is a release-blocker for other projects.

File: docs/root/version_history/current.rst [hunk 1]
```diff
@@ -13,6 +13,8 @@ Bug Fixes
 ---------
 *Changes expected to improve the state of the world and are unlikely to have negative effects*
 
+* zipkin: fix timestamp serializaiton in annotations. A prior bug fix exposed an issue with timestamps being serialized as strings.
```
Comment (htuch) [inline docs/root/version_history/current.rst:16]: Nit: typo on serialization

File: source/extensions/tracers/zipkin/span_buffer.cc [hunk 2]
```diff
@@ -133,7 +133,7 @@ JsonV2Serializer::toListOfSpans(const Span& zipkin_span, Util::Replacements& rep
       auto* annotation_entry_fields = annotation_entry.mutable_fields();
       (*annotation_entry_fields)[ANNOTATION_VALUE] = ValueUtil::stringValue(annotation.value());
       (*annotation_entry_fields)[ANNOTATION_TIMESTAMP] =
-          Util::uint64Value(annotation.timestamp(), annotation.value(), replacements);
+          Util::uint64Value(annotation.timestamp(), ANNOTATION_TIMESTAMP, replacements);
```
Comment (htuch) [inline source/extensions/tracers/zipkin/span_buffer.cc:136]: @dio does this make sense?

**Review by htuch — COMMENTED** [review]
LGTM but I'd like @dio to verify behavior intent.

File: docs/root/version_history/current.rst (hunk 1 above)
Comment (douglas-reid) [reply docs/root/version_history/current.rst]: fixed.

File: source/extensions/tracers/zipkin/span_buffer.cc (hunk 2 above)
Comment (dio) [reply source/extensions/tracers/zipkin/span_buffer.cc:136]: I think this is right. Sorry for the delay.

**Review by dio — DISMISSED** [review]
Thanks, @douglas-reid sorry for the delay. LGTM. If you fix the DCO, I'll re-approve this.

**Comment by htuch** [comment]
@douglas-reid please fix DCO and we can merge!

**Comment by douglas-reid** [comment]
> @douglas-reid please fix DCO and we can merge!

well, trying to fix the DCO has certainly cause quite a mess. please hold while I try to figure out what is happening.

**Comment by douglas-reid** [comment]
Trying again.

**Comment by douglas-reid** [comment]
@htuch @dio had to move to a new PR because the DCO cluster. please accept my apologies. please review / approve https://github.com/envoyproxy/envoy/pull/16072.

**Comment by douglas-reid** [comment]
Sorry, https://github.com/envoyproxy/envoy/pull/16073.  Today is not my day typing-wise.

**Pull request closed without merging**